### ISSUE TITLE
Matmul Ops

### ISSUE THREAD

**Issue by TheAiEngineer** [AiEngineering/tf2-ConvTasNet#14]
In `call()` of `ConvTasNet`, an InvalidArgumentError occured when creating `masked`

```py
    def call(self, signals: tf.Tensor):

        if signals.ndim != 3 or signals.shape[-1] != 1:
            raise ValueError(
                f"Expected 3D tensor (batch, channels==1, frame). Found {signals.shape}"
            )

        padded, num_pads = self.pad(signals)
        batch_size, num_padded_frames = padded.shape[2], padded.shape[2]
        feats = self.encoder(padded)
        masked = self.mask_generator(feats) * tf.expand_dims(signals, 1)
```

The trace is shown below:

```py
Exception has occurred: InvalidArgumentError
Incompatible shapes: [2,2,975104] vs. [2,1,121920,1] [Op:Mul]

During handling of the above exception, another exception occurred:

  File "/Users/justin/MSAI/AI-ML/CS-590/tf2-ConvTasNet/model/algorithm.py", line 253, in call
    masked = self.mask_generator(feats) * tf.expand_dims(signals, 1)
  File "/Users/justin/MSAI/AI-ML/CS-590/tf2-ConvTasNet/model/algorithm.py", line 295, in <module>
    zed = model.call(sigs[0])
```

**Comment by TheAiEngineer** [comment]
error is likely a result of typo at

```py
batch_size, num_padded_frames = padded.shape[2], padded.shape[2]
```

The objects are not supposed to be set to the same shape value.

**Comment by TheAiEngineer** [comment]
expand and masked are incompatible for matmul ops

![Screen Shot 2022-01-26 at 3 40 51 PM](https://user-images.githubusercontent.com/26209687/151243687-6344d3ba-d6e7-4092-97cd-d9a278a684b3.png)

**Comment by TheAiEngineer** [comment]
Continuing to troubleshoot dimension issues as data flows through network

![Screen Shot 2022-02-01 at 4 11 17 PM](https://user-images.githubusercontent.com/26209687/152052143-cd9dd7d6-ba2f-4297-b18f-c7ab75b771b7.png)

**Comment by TheAiEngineer** [comment]
When developing locally on CPU only, it is important to pass tensors to layers and models as 'channels last'. The explanation in `data_format` of [tf.keras.layers.SeparableConv1D](https://www.tensorflow.org/api_docs/python/tf/keras/layers/SeparableConv1D) provides details shown below
![Screen Shot 2022-02-04 at 10 20 23 AM](https://user-images.githubusercontent.com/26209687/152554700-1378b816-7da3-46ed-b5b1-6b5865543c4e.png)

**Comment by TheAiEngineer** [comment]
When a layer requires explicit padding prior to operations in tensorflow and keras, it is a must to 'pre-pad' the input tensor with [ZeroPadding1D](https://www.tensorflow.org/api_docs/python/tf/keras/layers/ZeroPadding1D), [ZeroPadding3D](https://www.tensorflow.org/api_docs/python/tf/keras/layers/ZeroPadding3D), [ZeroPadding3D](https://www.tensorflow.org/api_docs/python/tf/keras/layers/ZeroPadding3D).

An example is shown below in the highlighted `self.pre_pad` codelines.

_Create the padding layer_

![Screen Shot 2022-02-04 at 10 31 06 AM](https://user-images.githubusercontent.com/26209687/152556513-00776ded-7aa3-4345-9027-84279cc3ee07.png)


_Use the padding layer_
![Screen Shot 2022-02-04 at 10 29 44 AM](https://user-images.githubusercontent.com/26209687/152556176-cb9558b9-afa6-4fe8-a484-0735fca19c04.png)
